package/kernel/linux/modules/netfilter.mk

```diff
@@ -1190,3 +1190,14 @@ define KernelPackage/nft-compat
 endef
 
 $(eval $(call KernelPackage,nft-compat))
+
+define KernelPackage/nft-tproxy
+  SUBMENU:=$(NF_MENU)
+  TITLE:=Netfilter nf_tables tproxy support
+  DEPENDS:=+kmod-nft-core kmod-ipt-tproxy
```

Comment:
Based on:
https://cateee.net/lkddb/web-lkddb/NFT_TPROXY.html
There is no dependency on ipt-tproxy, why did you add that? I believe it should be the other way around.

Reply:
```
config NFT_TPROXY
	tristate "Netfilter nf_tables tproxy support"
	depends on IPV6 || IPV6=n
	select NF_DEFRAG_IPV4
	select NF_DEFRAG_IPV6 if NF_TABLES_IPV6
	select NF_TPROXY_IPV4
	select NF_TPROXY_IPV6 if NF_TABLES_IPV6
	help
	  This makes transparent proxy support available in nftables.
```
Kconfig shows that NFT_TPROXY depends on: NF_DEFRAG_IPV4, NF_DEFRAG_IPV6, NF_TPROXY_IPV4, NF_TPROXY_IPV6.
```
# tproxy

$(eval $(call nf_add,IPT_TPROXY,CONFIG_NETFILTER_XT_MATCH_SOCKET, $(P_XT)xt_socket))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV4, $(P_V4)nf_socket_ipv4))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_SOCKET_IPV6, $(P_V6)nf_socket_ipv6))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NETFILTER_XT_TARGET_TPROXY, $(P_XT)xt_TPROXY))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_TPROXY_IPV4, $(P_V4)nf_tproxy_ipv4))
$(eval $(call nf_add,IPT_TPROXY,CONFIG_NF_TPROXY_IPV6, $(P_V6)nf_tproxy_ipv6))
```
Then I found them in openwrt's IPT_TPROXY definition. So I added ipt-tproxy to the nft-tproxy dependency.

Reply:
```
```
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_DEFRAG_IPV6, $(P_V6)nf_defrag_ipv6),))
```
# kernel only
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_CONNTRACK, $(P_XT)nf_conntrack),))
$(eval $(if $(NF_KMOD),$(call nf_add,NF_CONNTRACK,CONFIG_NF_DEFRAG_IPV4, $(P_V4)nf_defrag_ipv4),))
```
Addendum: ipt-tproxy depends on ipt-conntrack, and ipt-conntrack depends on nf-conntrack.

Reply:
Yes, but that dependency pulls in iptables/xtables which is to be dropped soon.
I believe you should add those configs that you need to the nf-tproxy package you're adding and removing them from the ipt-tproxy, then you make ipt-tproxy depend on nf-tproxy.
Tomorrow when I'm at the computer I can give you some more pointer if you need, but I believe you get it.
Just out of curiosity, do you need this for some package?

Reply:
Yes, but that dependency pulls in iptables/xtables which is to be dropped soon.
I believe you should add those configs that you need to the nf-tproxy package you're adding and removing them from the ipt-tproxy, then you make ipt-tproxy depend on nf-tproxy.
Tomorrow when I'm at the computer I can give you some more pointer if you need, but I believe you get it.
Just out of curiosity, do you need this for some package?

Update to your addendum:
Yeah, you should chenge the dependency to nf_conntrack and that way you pull in what you need without selecting iptables

Reply:
> Just out of curiosity, do you need this for some package?

I'm a developer coming from China and the Chinese network you know, can't even access Google, which greatly affects my work.
PS: Baidu never returns the results I want.

In the past, I have used iptables tproxy on my router to redirect packets to proxy. After recently upgrading to the latest openwrt, I found that nft replaced iptables, but nft's tproxy doesn't work. Now I can only use nat to redirect ipv4 tcp packets, ipv6 and udp don't work.

Then I asked for help on the openwrt forum, but no one replied to me.
[https://forum.openwrt.org/t/how-to-make-nft-support-tproxy/118518](url)
[https://forum.openwrt.org/t/request-to-compile-nft-tproxy-ko-to-make-nft-support-tproxy/119974](url)

So I had to learn how to compile openwrt and how to add packages to openwrt by myself.

Reply:
Thanks for the suggestion, I've readjusted the package dependencies, please review again.

PS: I found that the first log was not written according to the specification, so I reworked it.

Reply:
Yeah, I believe no one uses these modules, but your work is greatly appreciated!

Reply:
Yeah, I believe no one uses these modules, but your work is greatly appreciated!